[mlir][Bufferization]Support to_tensor / to_buffer in One-Shot Bufferize analysis

kimm240 opened the pull request:
Summary
-------

This patch teaches One-Shot Bufferize how to reason about
`bufferization.to_tensor` and `bufferization.to_buffer` without crashing,
and relaxes the previous restriction that rejected `to_tensor` ops without
`restrict`.

The main ideas are:
- Treat `ToTensorOp` / `ToBufferOp` as proper BufferizableOpInterface ops
  with well-defined aliasing and read/write semantics.
- Keep tensor-only alias analysis strictly on tensor values.
- Model memref ↔ tensor alias sets for `to_tensor` / `to_buffer` explicitly
  in OneShotAnalysis, instead of trying to push memrefs through the generic
  tensor alias helpers.


Changes
-------

1) BufferizableOpInterface semantics for ToTensorOp / ToBufferOp

- `ToTensorOp`:
  - Implements `getAliasingValues`, `bufferizesToMemoryRead`,
    `bufferizesToMemoryWrite` in TableGen.
  - The tensor result is modeled as aliasing the memref operand.
  - Reads from the memref but does not write.

- `ToBufferOp`:
  - Implements `bufferizesToMemoryRead`, `bufferizesToMemoryWrite`,
    and `resultBufferizesToMemoryWrite` in a conservative way:
    - The op itself does not perform a write; writes are modeled solely
      on downstream memref users.
  - `getAliasingValues` returns an empty list from the tensor-analysis
    point of view; the memref result is not treated as a tensor alias.


2) OneShotAnalysis alias handling for to_tensor / to_buffer

- In `checkPreBufferizationAssumptions`:
  - Add explicit alias-set unions for `ToTensorOp` and `ToBufferOp`:
    - `to_tensor`: union the tensor result with the memref operand,
      and propagate existing aliases of the memref.
    - `to_buffer`: union the memref result with the tensor operand,
      and propagate existing aliases of the tensor.
- This keeps memref/tensor aliasing localized to OneShotAnalysis where
  alias sets are explicitly managed, instead of going through the
  tensor-only BufferizableOpInterface helpers.


3) Restrict alias helpers to tensor values

- `AnalysisState::getAliasingOpOperands(Value)`:
  - Now early-returns an empty list for non-tensor values.
  - This prevents tensor-only alias analysis from trying to follow
    aliases starting at memref results (e.g., `ToBufferOp`), which
    previously led to assertions in the one-shot pipeline.


Tests
-----

- `mlir/test/Dialect/Bufferization/Transforms/one-shot-module-bufferize-invalid.mlir`:
  - Replace the previous "to_tensor without `restrict` must error" tests
    with positive tests that verify `to_tensor` without `restrict`
    now bufferizes successfully.

- `mlir/test/Dialect/Bufferization/Transforms/one-shot-module-bufferize.mlir`:
  - `test_to_tensor_without_restrict_works`:
    - Checks that `to_tensor` without `restrict` is lowered to a plain
      `memref.load` without extra alloc/copy.
  - `forum_to_tensor_to_buffer_example` (renamed/commented generically):
    - `tensor<2xf32> -> to_buffer -> external memref call ->
       to_tensor -> tensor use`
    - Ensures this pattern no longer crashes the one-shot bufferization
      pipeline and that the external memref function remains memref-based.


Notes
-----

- This patch focuses on making One-Shot Bufferize robust and correct
  for the common `to_tensor` / `to_buffer` patterns (including the
  external-call example), while keeping the analysis conservative.
- More aggressive alias reasoning through `to_buffer` could be added
  in follow-up work if needed, but is intentionally out of scope here.
  
  Background / Reference
----------------------

- This work was motivated by the crash scenario discussed on the MLIR
  discourse thread:

  [One-Shot Bufferizer: Bufferization fails in the presence of
  `bufferization.to_memref` and `bufferization.to_tensor`][1]

  [1]: https://discourse.llvm.org/t/one-shot-bufferizer-bufferization-fails-in-the-presence-of-bufferization-to-memref-and-bufferization-to-tensor/62211

Commit:
Add tests for to_tensor/to_buffer op handling issues

- Add test case for to_tensor without restrict attribute that returns a tensor
  This test verifies that to_tensor ops without restrict cause an error
  as expected by One-Shot Analysis (OneShotAnalysis.cpp:1208-1217)

- Add test case for to_buffer copy creation
  This test verifies that to_buffer ops create a copy because they cannot
  be analyzed (one-shot-module-bufferize.mlir:696-702)

These tests reproduce the to_tensor/to_memref op handling issues from PR1.

Commit:
Enable to_tensor/to_buffer alias analysis in one-shot bufferize

- Implement BufferizableOpInterface behavior for ToTensorOp/ToBufferOp in TableGen
  (aliasing, read/write semantics)
- Set up explicit alias unions for to_tensor/to_buffer in OneShotAnalysis
- Remove restriction that rejected to_tensor without restrict and update tests
  to cover the new behavior

Commit:
Cover forum to_tensor/to_buffer example in tests

Add a one-shot bufferization test that mirrors the forum IR example
with to_buffer (to_memref), an external memref function call, and a
subsequent to_tensor use. This ensures that our alias handling and
crash behavior can be validated against the real-world scenario that
motivated this change.

Also document the use of const_cast in ToTensorOp/ToBufferOp
BufferizableOpInterface implementations so that the rationale is clear
in the generated C++ code.

Commit:
Restrict alias analysis helpers to tensor values for ToBufferOp

Teach AnalysisState::getAliasingOpOperands to early-return for
non-tensor values so that tensor-only alias analysis does not try to
follow aliases through memref results such as ToBufferOp. This keeps
non-tensor aliasing (e.g., between ToBufferOp results and their tensor
operands) in OneShotAnalysis where alias sets are explicitly unioned.

Also update ToBufferOp's BufferizableOpInterface implementation so that
neither its operand nor its result are treated as implicit memory
writes; writes are modeled solely on downstream memref users.

Commit:
Clarify to_buffer/to_tensor external-call test description

Update the comment around the to_buffer + external memref call +
to_tensor test to describe the scenario in generic terms instead of
referencing the original forum thread. The test now simply documents
that this pattern must not crash the one-shot bufferization pipeline
and keeps the IR example self-contained.

matthias-springer reviewed (commented):
How does this work in the presence of memref ops that do not implement the `BufferizableOpInterface`. E.g.:

```mlir
%0 = bufferization.to_memref %t
%1 = memref.subview %0
%2 = memref.subview %1
%3 = bufferization.to_tensor %2
```

How does the bufferization analysis know that %0 and %3 are aliases?

matthias-springer reviewed (commented):
There are a few issues with this PR. E.g., this snippet no longer bufferizes correctly. There used to be a `memref.copy`, which is needed for correctness, but it is now dropped.

```
~/llvm-project$ cat a.mlir
func.func @foo() {
  %0 = tensor.empty() : tensor<10xf32>
  %cst = arith.constant 5.0 : f32
  %cst2 = arith.constant 6.0 : f32
  %c0 = arith.constant 0 : index
  %2 = linalg.fill ins(%cst : f32) outs(%0 : tensor<10xf32>) -> tensor<10xf32>
  %3 = bufferization.to_buffer %2 : tensor<10xf32> to memref<10xf32>
  memref.store %cst2, %3[%c0] : memref<10xf32>
  %4 = tensor.extract %2 [%c0] : tensor<10xf32>
  vector.print %4 : f32
  return
}

~/llvm-project$ build/bin/mlir-opt a.mlir -one-shot-bufferize
module {
  func.func @foo() {
    %alloc = memref.alloc() {alignment = 64 : i64} : memref<10xf32>
    %cst = arith.constant 5.000000e+00 : f32
    %cst_0 = arith.constant 6.000000e+00 : f32
    %c0 = arith.constant 0 : index
    linalg.fill ins(%cst : f32) outs(%alloc : memref<10xf32>)
    memref.store %cst_0, %alloc[%c0] : memref<10xf32>
    %0 = memref.load %alloc[%c0] : memref<10xf32>
    vector.print %0 : f32
    return
  }
}
```

The reason in this case is that `ToBufferOp::bufferizesToMemoryWrite` now returns `false`.

I think there are similar issues with `ToTensorOp`. In the absence of a memref analysis, it is not safe to allow such ops without `restrict`.

Commit:
Fix ToBufferOp to correctly handle memory writes for copy insertion

This fixes the copy omission issue identified by the reviewer where
memref.copy operations were missing after bufferization.to_buffer,
leading to incorrect values being loaded and violating tensor SSA
immutability.

Changes:
- ToBufferOp::bufferizesToMemoryWrite now returns !getReadOnly() instead
  of false, ensuring that when read_only is not set, the operation is
  treated as a memory write, which triggers proper copy insertion.
- Added ToBufferOp::resultBufferizesToMemoryWrite override to handle
  memref results correctly, avoiding assertion failures when the default
  implementation tries to call getAliasingOpOperands on memref types.
- ToBufferOp::getAliasingValues now returns the input tensor to enable
  RaW conflict detection when the result memref is written to and the
  input tensor is read from.
- Fixed test CHECK pattern to match actual output format.

This ensures that when a new tensor (e.g., from linalg.fill) is
converted to a memref via to_buffer, the memref does not share memory
with the original tensor buffer, maintaining tensor immutability.

Commit:
Add test case for linalg.fill + to_buffer copy insertion

This adds a test case based on the reviewer's specific IR example that
demonstrates the fix for copy omission. The test verifies that when a
new tensor (from linalg.fill) is converted to a memref via to_buffer,
and then the memref is written to while the original tensor is read
from, a copy is correctly inserted to maintain tensor immutability.

The test case:
- Creates a tensor with tensor.empty()
- Fills it with linalg.fill
- Converts to memref with to_buffer
- Writes to the memref
- Reads from the original tensor

This ensures memref.copy is inserted between the fill and the store,
maintaining the immutability of the original tensor.

Commit:
Refine writable-buffer analysis for to_tensor/to_buffer aliasing

- Restore explicit aliasing between to_tensor/to_buffer operands and results
  so that RaW conflicts through these ops are still detected.
- Restrict non-writable buffer detection to tensor values only, avoiding
  spurious NOT-WRITABLE annotations on memrefs (e.g., memref.alloc).
- This keeps One-Shot Analysis tensor-centric while still modeling the
  necessary tensor/memref aliasing for conflict detection and copy insertion.
